### PR TITLE
test(client): add `vitess` flavor in relationMode test suite

### PR DESCRIPTION
**Pull request by jkomyno**
This PR:
- adds support for `vitess-8` (using the Docker image `vitess/vttestserver:mysql80`)
- introduces the concept of `ProviderFlavor`, a superset of `Providers` that allows distinguishing slightly different variants of providers. E.g., the `vitess-8` flavor uses `mysql` as the `provider` in the Prisma schema, but its url is mapped to the `TEST_FUNCTIONAL_VITESS_8_URI` env variable rather than `TEST_FUNCTIONAL_MYSQL_URI`.

**Note**: I have noticed Vitess is particularly slow. Maybe there's something sketchy on my local machine and it's faster on the CI.

Related https://github.com/prisma/prisma/issues/6927.

### CODE REVIEW

File: packages/client/tests/functional/relationMode/tests_1-to-1.ts [hunk 1]
```diff
@@ -601,6 +603,7 @@ testMatrix.setupTestSuite(
                           [Providers.MYSQL]: 'Unique constraint failed on the constraint: `ProfileOneToOne_userId_key`',
                           [Providers.SQLSERVER]: 'Unique constraint failed on the constraint: `dbo.ProfileOneToOne`',
                           [Providers.SQLITE]: 'Unique constraint failed on the fields: (`userId`)',
+                          [ProviderFlavors.VITESS_8]: 'Unique constraint failed on the (not available)',
```
Comment (jkomyno) [inline packages/client/tests/functional/relationMode/tests_1-to-1.ts:606]: Btw, notice that Prisma error messages on Vitess are somewhat lacking compared to the other providers.

File: .buildkite/publish/docker-compose.yml [hunk 2]
```diff
@@ -125,6 +127,20 @@ services:
     ports:
       - '3307:3306'
 
+  vitess-8:
+    image: vitess/vttestserver:mysql80@sha256:9412e3d51bde38e09c3039090b5c68808e299579f12c79178a4ec316f7831889
```
Comment (Jolg42) [inline .buildkite/publish/docker-compose.yml:131]: Note: it's the same version we test in prisma-engines

File: packages/client/tests/functional/relationMode/tests_1-to-1.ts (hunk 1 above)
Comment (janpio) [reply packages/client/tests/functional/relationMode/tests_1-to-1.ts:606]: https://github.com/prisma/prisma/issues/10829

File: packages/client/tests/functional/_utils/relationMode/computeMatrix.ts [hunk 3]
```diff
@@ -42,9 +46,10 @@ export function computeMatrix({ relationMode, providersDenyList }: ComputeMatrix
   //
   // We skip these combinations in the matrix (= filtering them out)
 
-  const referentialActionsDenylistByProvider = {
+  const referentialActionsDenylistByProviderFlavor = {
     foreignKeys: {
       [Providers.SQLSERVER]: ['Restrict'],
+      [ProviderFlavors.VITESS_8]: referentialActionsBase, // no action is executed for Vitess & relationMode="foreignKeys"
```
Comment (Jolg42) [inline packages/client/tests/functional/_utils/relationMode/computeMatrix.ts:52]: It's confusing because it can be read as `NO ACTION` which is an action 😄 
```suggestion
      [ProviderFlavors.VITESS_8]: referentialActionsBase, // skip all actions for Vitess & relationMode="foreignKeys"
```

Comment (Jolg42) [reply packages/client/tests/functional/_utils/relationMode/computeMatrix.ts:52]: But Why do we need to skip all?

Comment (Jolg42) [reply packages/client/tests/functional/_utils/relationMode/computeMatrix.ts:52]: I improved the comment

**Review by Jolg42 — APPROVED** [review]
Looks good!

Note: only the `relationMode` tests are also tested on vitess.

**Comment by janpio** [comment]
> Note: I have noticed Vitess is particularly slow. Maybe there's something sketchy on my local machine and it's faster on the CI.

What is the impact on CI/test run time?

**Comment by jkomyno** [comment]
> > Note: I have noticed Vitess is particularly slow. Maybe there's something sketchy on my local machine and it's faster on the CI.
> 
> What is the impact on CI/test run time?

You can see that `CI / client-relationMode (prisma, ubuntu-latest, 16)` got canceled after 35m. I will investigate ways of reducing the time needed for that.

File: docker/docker-compose.yml [hunk 4]
```diff
@@ -38,14 +38,21 @@ services:
...
       PORT: '33804' # unused in testing, but required by vttestserver
-      KEYSPACES: 'keyspace' # unused in testing, but required by vttestserver
+      KEYSPACES: 'unsharded' # unused in testing, but required by vttestserver
       NUM_SHARDS: '1' # unused in testing, but required by vttestserver
-
       MYSQL_BIND_HOST: '0.0.0.0'
       FOREIGN_KEY_MODE: 'disallow'
       VT_DIALECT: 'mysql80'
+      PLANNER_VERSION: 'gen4fallback'
+      TOPOLOGY_FLAGS: '-shutdown_grace_period 2'
```
Comment (Jolg42) [inline docker/docker-compose.yml:55]: What are these settings?

Comment (jkomyno) [reply docker/docker-compose.yml:55]: - `PLANNER_VERSION`: it uses the most recent planner for Vitess, allowing a fallback to the previous planner version in case the newest fails (see the [docs](https://vitess.io/docs/15.0/get-started/vttestserver-docker-image/#environment-variables))
- `TOPOLOGY_FLAGS`: added `-shutdown_grace_period 2` as suggested in [this comment](https://github.com/vitessio/vitess/issues/9936#issuecomment-1074498933), using an environment variable shown in [here](https://vitess.io/docs/13.0/reference/programs/vttablet/#managed-mysql) (which, as far as I can tell, is overridable)

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -253,7 +258,7 @@ jobs:
         run: |
           echo "PRISMA_CLIENT_ENGINE_TYPE=${{ matrix.queryEngine }}" >> $GITHUB_ENV
 
-      - run: docker-compose -f docker/docker-compose.yml up --detach postgres postgres_isolated mysql mysql_isolated mssql mongo cockroachdb
+      - run: docker-compose -f docker/docker-compose.yml up --detach postgres postgres_isolated mysql mysql_isolated mssql mongo cockroachdb vitess-8
```
Comment (janpio) [inline .github/workflows/test.yml:261]: Is this needed?

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -310,7 +315,7 @@ jobs:
           username: ${{ secrets.DOCKERHUB_USERNAME }}
           password: ${{ secrets.DOCKERHUB_TOKEN }}
 
-      - run: docker-compose -f docker/docker-compose.yml up --detach postgres mysql mssql mongo cockroachdb
+      - run: docker-compose -f docker/docker-compose.yml up --detach postgres mysql mssql mongo cockroachdb vitess-8
```
Comment (janpio) [inline .github/workflows/test.yml:318]: Is this needed?

File: docker/docker-compose.yml (hunk 4 above)
Comment (Jolg42) [reply docker/docker-compose.yml:55]: If these don't change anything (does it make the test faster or anything?), maybe we could remove them. Going with the defaults feels better, I think.

File: .github/workflows/test.yml (hunk 5 above)
Comment (Jolg42) [reply .github/workflows/test.yml:261]: I don't think it's needed now because we have no tests in client functional tests that test on vitess. But as soon as we add one test, it will be needed.

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -1033,7 +1040,7 @@ jobs:
           CI: true
           SKIP_GIT: true
           GITHUB_CONTEXT: ${{ toJson(github) }}
-          NODE_OPTIONS: '--max-old-space-size=8096'
+          NODE_OPTIONS: '--max-old-space-size=10240' # allow Node.js to allocate at most 10gb of heap
```
Comment (Jolg42) [inline .github/workflows/test.yml:1043]: This is about the `CLIENT (memory tests)` job, so, it looks unrelated to the current PR. What is it solving?

Comment (jkomyno) [reply .github/workflows/test.yml:1043]: Random flakiness on Windows that was happening on this PR (although not caused by this PR)

File: docker/docker-compose.yml (hunk 4 above)
Comment (jkomyno) [reply docker/docker-compose.yml]: It was the latest attempt to make Vitess faster. Removing them now, let's see whether the test suite stays within the new timeout boundaries (60 mins)

File: .github/workflows/test.yml (hunk 5 above)
Comment (jkomyno) [reply .github/workflows/test.yml]: So @Jolg42, is the `relationMode` functional test suite excluded from the rest of the client functional tests unless `--relation-mode-tests-only` is specified? Or is it run regardless, but with the implicit `relationMode = "foreignKeys"` setting?

Comment (Jolg42) [reply .github/workflows/test.yml]: `relationMode` tests are always ignored, unless `--relation-mode-tests-only` is passed, which we only do in the client relationMode job on GitHub Actions

File: .github/workflows/test.yml (hunk 6 above)
Comment (jkomyno) [reply .github/workflows/test.yml]: No, removed.

File: .github/workflows/test.yml (hunk 5 above)
Comment (jkomyno) [reply .github/workflows/test.yml]: Removed.

File: .buildkite/test/docker-compose.14.yml [hunk 8]
```diff
@@ -64,6 +65,7 @@ services:
       - mongodb_migrate
       - mongodb_migrate_seed
       - cockroachdb
+      - vitess-8
```
Comment (Jolg42) [inline .buildkite/test/docker-compose.14.yml:68]:
```suggestion
# TODO: uncomment when we test vitess in client functional test suite
# Currently we only test `relationMode` tests in GitHub Actions
#      - vitess-8
```

File: .buildkite/publish/docker-compose.yml [hunk 9]
```diff
@@ -74,6 +75,7 @@ services:
       - mongodb_migrate
       - mongodb_migrate_seed
       - cockroachdb
+      - vitess-8
```
Comment (Jolg42) [inline .buildkite/publish/docker-compose.yml:78]:
```suggestion
# TODO: uncomment when we test vitess in client functional test suite
# Currently we only test `relationMode` tests in GitHub Actions
#      - vitess-8
```
